packages/ai-jsx/src/core/completion.tsx

```diff
@@ -40,18 +42,56 @@ export type ModelComponent<T extends ModelPropsWithChildren> = Component<T>;
  */
 export interface FunctionDefinition {
   description?: string;
-  parameters: Record<string, FunctionParameter>;
+  parameters: FunctionParameters;
+}
+
+export function getParametersSchema(parameters: FunctionParameters) {
+  if (parameters instanceof z.Schema) {
+    const jsonSchema = zodToJsonSchema(parameters);
+    if (!('type' in jsonSchema) || jsonSchema.type !== 'object') {
+      throw new AIJSXError(
+        `Function parameters param must be an object, but was: ${JSON.stringify(jsonSchema, null, 2)}`,
```

Comment:
It's a bit confusing to say "input must be an object" when the input `{a: true}` would trigger this error.

Reply:
I think the current form is a footgun, because things like this don't work:

```tsx
const Elements = z.union([Element, z.array(Element)]);

<JsonChatCompletion schema={Elements} retries={1}>
```

OpenAI requires that the top-level type be an object. However, it seems more natural in Zod to pass in any Zod type.

The way I've been solving this manually is to wrap with an outer object:

```tsx
  const Elements = z.union([Element, z.array(Element)]);
  const elementsWrapper = z.object({root: Elements });
<JsonChatCompletion schema={elementsWrapper} retries={1}>
```

But this is kind of annoying for me and will be confusing for newcomers.

Maybe the `JsonChatCompletion` component should do the wrapping / unwrapping automatically? I'm slightly concerned it would be too magical, but I think on net it's worth eliminating the footgun.

At the very least, we could update the error message to tell the dev exactly what they need to do.

Reply:
If we can make it work every time that's good, but sometimes the user might describe the JSON in their prompt (since it doesn't usually follow the schema too well). In that case the difference between user's view and the actual schema can cause confusion.

Reply:
If we do that, we'll have to come up with a generic name for it, but names matter with LLMs and it can be annoying for users who don't want this extra step.

This is kind of an issue with OpenAI, not us, so maybe we don't need to address it? Maybe they'll fix it, or another provider will come up with a better interface. Handling this will put the burden on us to make sure it is the best thing to do.

Reply:
> If we can make it work every time that's good, but sometimes the user might describe the JSON in their prompt (since it doesn't usually follow the schema too well).

Great point about the difference with the prompt itself causing a problem.

But is it still true that the model doesn't follow the schema well with function calls? I thought OpenAI was claiming that it did.

Reply:
> This is kind of an issue with OpenAI, not us

I disagree. OpenAI says, "give us any JSON schema that's an object", and that's easy to follow.

We say, "give us any Zod schema", but then at runtime, Zod schemas that aren't objects at the top-level fail. And we can see how counter-intuitive this is, because both you and I wrote Zod schemas that aren't objects at the top level in these PRs. :smile:

But I agree with you that it's bad if the user does their own prompt engineering and then the precise form doesn't match. So for now, let's just make an explicit note in the docs.

Reply:
Oh I agree that we should make it a compile time error and not a runtime one. That's why I asked you before. I'm gonna try your idea here now (sorry got called into a meeting):
https://www.typescriptlang.org/play?ssl=1&ssc=1&pln=3&pc=1#code/JYWwDg9gTgLgBALzgMyhEcDkCIBNMDcAUEQMYQB2AzvAIYUCeA8gEYBWApqTAMqkAWHELTgBeRADooXaLgAUCCQFcKAawoQA7hTkBKXcRgMwHOABUxk4BWQcoAHiMmIyOPWbsuvAUNoA+EnJqeAgALnNLAG9kCDCsFlooTABfYiIAenS4AHVTTXp4GH5gKjgYCBRaYAAbMkoaOAgAJnCLcRgoJQ4CIA

Reply:
As per discussion, this was made to be more strict by enforcing a `z.ZodObject` instead of a general `z.Schema`.

Reply:
I felt good about where we ended up on this, but I ran into another related footgun to this.

I made a component:

```tsx
          <JsonChatCompletion>
            <UserMessage>Generate some information about hotels, including each hotel's name and how many stars it is</UserMessage>
          </JsonChatCompletion>
```

It generated this JSON:

```json
[{"hotel_name":"Hotel A","stars":4},{"hotel_name":"Hotel B","stars":3},{"hotel_name":"Hotel C","stars":5}]
```

This is valid for my usecase, but JCC threw a validation error:

```
    validationError: '[\n' +
      '  {\n' +
      '    "code": "invalid_type",\n' +
      '    "expected": "object",\n' +
      '    "received": "array",\n' +
      '    "path": [],\n' +
      '    "message": "Expected object, received array"\n' +
      '  }\n' +
      ']'
```

I believe this is because if `schema` is not passed, JCC defaults to `object`, so this array fails validation:

https://github.com/fixie-ai/ai-jsx/blob/55ddca6efa8999c8b0632274a811bba82ebc3986/packages/ai-jsx/src/batteries/constrained-output.tsx#L99.

Do you have thoughts as to what a good solution here is?

Reply:
As a workaround, I can manually specify a `schema` arg that's an object. But this would be confusing to people.

Reply:
Hmm yeah good point.

Q: do you want JCCFC specifically? If yes, then the top-level has to be an object (unless we do wrapping/unwrapping which I'm against and could add to confusion).

If not, then we can go back to not using JCCFC if the schema is not provided. I agree that this is a bit weird: presence or not of `schema` shouldn't change the underlying mechanism without any warnings.

Maybe the solution is to expose both and let people decide, but I think you were against that. We might also make it more explicit by having an option that says `useFunctionCall = ['true', 'false', 'auto']` (or something like that).